File: preflight-checks/dcgm-diag/pkg/health/health.go
// Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Package health provides health event reporting functionality.
package health

import (
	"context"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"os"
	"strings"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/timestamppb"
	"k8s.io/apimachinery/pkg/util/wait"

	pb "github.com/nvidia/nvsentinel/data-models/pkg/protos"
)

const (
	agentName      = "preflight-dcgm-diag"
	componentClass = "GPU"
	checkName      = "DCGM_DIAGNOSTIC"

	maxRetries  = 5
	retryDelay  = 2 * time.Second
	retryFactor = 1.5
	retryJitter = 0.1
	sendTimeout = 10 * time.Second
)

// SendHealthEvent sends a health event to the platform connector.
func SendHealthEvent(socketPath string, gpuUUIDs []string, isHealthy, isFatal bool, message string) error {
	event := buildHealthEvent(gpuUUIDs, isHealthy, isFatal, message)
	healthEvents := &pb.HealthEvents{
		Version: 1,
		Events:  []*pb.HealthEvent{event},
	}

	slog.Info("Sending health event",
		"isHealthy", isHealthy,
		"isFatal", isFatal,
		"gpuCount", len(gpuUUIDs),
		"message", message)

	// Handle unix:// prefix
	socketPath = strings.TrimPrefix(socketPath, "unix://")

	backoff := wait.Backoff{
		Steps:    maxRetries,
		Duration: retryDelay,
		Factor:   retryFactor,
		Jitter:   retryJitter,
	}

	err := wait.ExponentialBackoff(backoff, func() (bool, error) {
		sendErr := sendToConnector(socketPath, healthEvents)
		if sendErr == nil {
			slog.Info("Health event sent successfully")

			return true, nil
		}

		if isRetryableError(sendErr) {
			slog.Warn("Retryable error sending health event, will retry", "error", sendErr)

			return false, nil
		}

		slog.Error("Non-retryable error sending health event", "error", sendErr)

		return false, fmt.Errorf("non-retryable error: %w", sendErr)
	})
	if err != nil {
		return fmt.Errorf("failed to send health event after %d retries: %w", maxRetries, err)
	}

	return nil
}

func buildHealthEvent(gpuUUIDs []string, isHealthy, isFatal bool, message string) *pb.HealthEvent {
	entities := make([]*pb.Entity, 0, len(gpuUUIDs))
	for _, uuid := range gpuUUIDs {
		entities = append(entities, &pb.Entity{
			EntityType:  "GPU_UUID",
			EntityValue: uuid,
		})
	}

	nodeName := os.Getenv("NODE_NAME")
	if nodeName == "" {
		nodeName = "unknown"
	}

	// For healthy events, just store the result without triggering remediation
	recommendedAction := pb.RecommendedAction_RUN_FIELDDIAG
	processingStrategy := pb.ProcessingStrategy_EXECUTE_REMEDIATION

	if isHealthy {
		recommendedAction = pb.RecommendedAction_NONE
		processingStrategy = pb.ProcessingStrategy_STORE_ONLY
	}

	return &pb.HealthEvent{
		Version:            1,
		Agent:              agentName,
		ComponentClass:     componentClass,
		CheckName:          checkName,
		IsFatal:            isFatal,
		IsHealthy:          isHealthy,
		Message:            message,
		RecommendedAction:  recommendedAction,
		EntitiesImpacted:   entities,
		GeneratedTimestamp: timestamppb.Now(),
		NodeName:           nodeName,
		ProcessingStrategy: processingStrategy,
	}
}

func sendToConnector(socketPath string, healthEvents *pb.HealthEvents) error {
	ctx, cancel := context.WithTimeout(context.Background(), sendTimeout)
	defer cancel()

	conn, err := grpc.NewClient(
		fmt.Sprintf("unix://%s", socketPath),
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		return fmt.Errorf("failed to connect to platform connector: %w", err)
	}

	defer conn.Close()

	client := pb.NewPlatformConnectorClient(conn)

	_, err = client.HealthEventOccurredV1(ctx, healthEvents)
	if err != nil {
		return fmt.Errorf("failed to send health event: %w", err)
	}

	return nil
}

// isRetryableError determines if an error is retryable.
func isRetryableError(err error) bool {
	if s, ok := status.FromError(err); ok {
		if s.Code() == codes.Unavailable || s.Code() == codes.DeadlineExceeded {
			return true
		}
	}

	if _, ok := err.(interface{ Temporary() bool }); ok {
		return true
	}

	if errors.Is(err, io.EOF) {
		return true
	}

	errStr := err.Error()

	return strings.Contains(errStr, "connection reset by peer") ||
		strings.Contains(errStr, "broken pipe") ||
		strings.Contains(errStr, "connection refused")
}

File: preflight-checks/dcgm-diag/dcgm_diag/health.py
# Copyright (c) 2026, NVIDIA CORPORATION.  All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
from time import sleep

import grpc
from google.protobuf.timestamp_pb2 import Timestamp

from .protos import health_event_pb2 as pb
from .protos import health_event_pb2_grpc as pb_grpc

log = logging.getLogger(__name__)

MAX_RETRIES = 5
INITIAL_DELAY = 2.0
BACKOFF_FACTOR = 1.5
RPC_TIMEOUT = 30.0


class HealthReporter:
    AGENT = "preflight-dcgm-diag"
    COMPONENT_CLASS = "GPU"
    CHECK_NAME = "DCGM_DIAGNOSTIC"

    def __init__(
        self,
        socket_path: str,
        node_name: str,
        processing_strategy: pb.ProcessingStrategy,
    ) -> None:
        self._socket_path = socket_path.removeprefix("unix://")
        self._node_name = node_name
        self._processing_strategy = processing_strategy

    def send_event(
        self,
        gpu_uuids: list[str],
        is_healthy: bool,
        is_fatal: bool,
        message: str,
    ) -> None:
        event = self._build_event(gpu_uuids, is_healthy, is_fatal, message)
        health_events = pb.HealthEvents(version=1, events=[event])

        log.info(
            f"Sending health event is_healthy={is_healthy} is_fatal={is_fatal} "
            f"gpu_count={len(gpu_uuids)} processing_strategy={self._processing_strategy} message={message}"
        )

        if not self._send_with_retries(health_events):
            raise RuntimeError(f"Failed to send health event after {MAX_RETRIES} retries")

    def _build_event(
        self,
        gpu_uuids: list[str],
        is_healthy: bool,
        is_fatal: bool,
        message: str,
    ) -> pb.HealthEvent:
        entities = [pb.Entity(entityType="GPU_UUID", entityValue=uuid) for uuid in gpu_uuids]

        recommended_action = pb.NONE if is_healthy else pb.CONTACT_SUPPORT

        timestamp = Timestamp()
        timestamp.GetCurrentTime()

        return pb.HealthEvent(
            version=1,
            agent=self.AGENT,
            componentClass=self.COMPONENT_CLASS,
            checkName=self.CHECK_NAME,
            isFatal=is_fatal,
            isHealthy=is_healthy,
            message=message,
            recommendedAction=recommended_action,
            entitiesImpacted=entities,
            generatedTimestamp=timestamp,
            nodeName=self._node_name,
            processingStrategy=self._processing_strategy,
        )

    def _send_with_retries(self, health_events: pb.HealthEvents) -> bool:
        delay = INITIAL_DELAY

        for attempt in range(MAX_RETRIES):
            try:
                with grpc.insecure_channel(f"unix://{self._socket_path}") as channel:
                    stub = pb_grpc.PlatformConnectorStub(channel)
                    stub.HealthEventOccurredV1(health_events, timeout=RPC_TIMEOUT)
                    log.info("Health event sent successfully")
                    return True
            except grpc.RpcError as e:
                log.warning(f"Failed to send health event (attempt {attempt + 1}/{MAX_RETRIES}): {e}")
                if attempt < MAX_RETRIES - 1:
                    sleep(delay)
                    delay *= BACKOFF_FACTOR

        return False
